xbmc/interfaces/python/PythonInvoker.cpp

```diff
@@ -367,36 +367,41 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
     onError(exceptionType, exceptionValue, exceptionTraceback);
   }
 
+  CSingleLock lock(m_critical);
   // no need to do anything else because the script has already stopped
   if (failed)
   {
     setState(stateToSet);
     return true;
   }
 
-  PyObject *m = PyImport_AddModule((char*)"xbmc");
-  if (m == NULL || PyObject_SetAttrString(m, (char*)"abortRequested", PyBool_FromLong(1)))
-    CLog::Log(LOGERROR, "CPythonInvoker(%d, %s): failed to set abortRequested", GetId(), m_sourceFile.c_str());
-
-  // make sure all sub threads have finished
-  for (PyThreadState* s = m_threadState->interp->tstate_head, *old = NULL; s;)
+  if (m_threadState)
   {
-    if (s == m_threadState)
-    {
-      s = s->next;
-      continue;
-    }
-    if (old != s)
+    PyObject *m = PyImport_AddModule((char*)"xbmc");
+    if (m == NULL || PyObject_SetAttrString(m, (char*)"abortRequested", PyBool_FromLong(1)))
+      CLog::Log(LOGERROR, "CPythonInvoker(%d, %s): failed to set abortRequested", GetId(), m_sourceFile.c_str());
+
+    // make sure all sub threads have finished
+    for (PyThreadState *s = m_threadState->interp->tstate_head, *old = nullptr; s != nullptr && m_threadState != nullptr;)
     {
-      CLog::Log(LOGINFO, "CPythonInvoker(%d, %s): waiting on thread %" PRIu64, GetId(), m_sourceFile.c_str(), (uint64_t)s->thread_id);
-      old = s;
-    }
+      for (; s && s == m_threadState;)
+        s = s->next;
 
-    CPyThreadState pyState;
-    Sleep(100);
-    pyState.Restore();
+      if (!s)
+        break;
 
-    s = m_threadState->interp->tstate_head;
+      if (old != s)
+      {
+        CLog::Log(LOGINFO, "CPythonInvoker(%d, %s): waiting on thread %" PRIu64, GetId(), m_sourceFile.c_str(), (uint64_t)s->thread_id);
+        old = s;
+      }
+
+      lock.Leave();
```

Comment:
While I'm not sure if anything here can throw, you should use an exit guard (CSingleExit)